### DISCUSSION TITLE
ROS2 Actions

### DISCUSSION THREAD

**Discussion by rokel** [eProsima/Integration-Service#148] [Ideas]
Is it possible to use SOSS with ROS2 actions? I think I understand that they are composed from regular messages and services so should be possible to use already, but is there a missing link needed for parsing `.action` files and generating the requisite plugins?

**Comment by mxgrey** [comment]
SOSS was primarily designed to help with interoperability between ROS and non-ROS middlewares. Since the concept of actions is somewhat unique to the ROS ecosystem, SOSS has not yet made an effort to support it. That being said, there may be value to having SOSS support it, since SOSS may be used to translate between ROS1 <--> ROS2 or between different DDS partitions in a ROS2 setup.

Support for actions could be added as an optional interface for the [`SystemHandle`](https://github.com/osrf/soss/blob/master/packages/core/include/soss/SystemHandle.hpp), alongside the optional publisher, subscriber, service, and client interfaces. Then any middleware that supports an action-like means of communication could implement those interfaces for its SOSS plugin.

All that being said, support for actions is not on the roadmap for the SOSS development team, so I wouldn't expect us to develop it any time soon. But we'd be happy to receive and review a contribution for this feature, and to provide guidance wherever it's feasible.
